nemo/collections/tts/torch/data_ssl_vocoder.py

```diff
@@ -0,0 +1,590 @@
+            mel_padded = torch.nn.functional.pad(mel, (0, max_mel_len - mel.size(1)), value=0)
+            mels_padded.append(mel_padded)
+
+        pitch_contours_padded = []
+        for pitch_contour in final_batch["pitch_contour"]:
+            pitch_contour_padded = torch.nn.functional.pad(
+                pitch_contour, (0, max_mel_len - pitch_contour.size(0)), value=0
+            )
+            pitch_contours_padded.append(pitch_contour_padded)
+
+        content_embeddings_padded = []
+        for encoded in final_batch["content_embedding"]:
+            encoded_padded = torch.nn.functional.pad(encoded, (0, max_encoded_len - encoded.size(1)), value=0)
+            content_embeddings_padded.append(encoded_padded)
+
+        durations_padded = []
+        for duration in final_batch["duration"]:
+            duration_padded = torch.nn.functional.pad(duration, (0, max_encoded_len - duration.size(0)), value=0.0)
+            durations_padded.append(duration_padded)
+
+        final_batch["audio"] = audios_padded
+        final_batch["mel_spectrogram"] = mels_padded
+        final_batch["pitch_contour"] = pitch_contours_padded
+        final_batch["content_embedding"] = content_embeddings_padded
+        final_batch["duration"] = durations_padded
+
+        for key in final_batch:
+            final_batch[key] = torch.stack(final_batch[key])
+
+        return final_batch
+
+    def compute_mean_speaker_embeddings(self, n_embeddings_per_speaker=100):
+        mean_speaker_embeddings = {}
+        speaker_counts = {}
+        for idx in range(len(self.data)):
+            sample = self.data[idx]
+            speaker = sample["speaker"]
+            if speaker in speaker_counts and speaker_counts[speaker] >= n_embeddings_per_speaker:
+                continue
+
+            rel_audio_path = Path(sample["audio_filepath"]).relative_to(self.base_data_dir).with_suffix("")
+            rel_audio_path_as_text_id = str(rel_audio_path).replace("/", "_")
+            speaker_emb_fn = f"speaker_embedding_{rel_audio_path_as_text_id}.pt"
+            speaker_emb_fp = os.path.join(self.sup_data_dir, speaker_emb_fn)
+            if os.path.exists(speaker_emb_fp):
+                embedding = torch.load(speaker_emb_fp)
+                if speaker not in mean_speaker_embeddings:
+                    mean_speaker_embeddings[speaker] = embedding
+                    speaker_counts[speaker] = 1
+                else:
+                    mean_speaker_embeddings[speaker] += embedding
+                    speaker_counts[speaker] += 1
+
+        for speaker in mean_speaker_embeddings:
+            mean_speaker_embeddings[speaker] /= speaker_counts[speaker]
+            l2_norm = torch.norm(mean_speaker_embeddings[speaker], p=2)
+            mean_speaker_embeddings[speaker] /= l2_norm
+
+        self.mean_speaker_embeddings = mean_speaker_embeddings
+
+    def filter_files(data, ignore_file, min_duration, max_duration, total_duration):
```

Comment:
Probably.. though the reason I made a separate class is that this one does not use text or any tokenizer etc. Also different from a VocoderDataset since the input is not mel-spectrogram. But I agree code duplication can still be avoided by reusing common functions.

Reply:
Hmm. I think it's somewhat closer to `TTSDataset` ("... used for training spectrogram generators and end-to-end TTS models ...") in terms of functionality, so it should be fine to subclass off of it rather than `VocoderDataset`... Just with a note that loads embeddings rather than text, so the processing for the input is different.

Actually, since this class is mainly paired with FastPitch (mel gen) rather than HiFi-GAN (vocoder) now--would it make sense to rename it to something that reflects that?

Either way they should share some function (either through inheritance or just separating it off into a helper function).

@blisc for viz and in case you have any opinions about classification here.

Reply:
It doesn't look like it subclasses anything here, and I think that's fine. Let's have this dataset be standalone for now, and once we want to further support this SSL pipeline, we can think about refactoring it in the future.